File: src/main/java/edu/missouristate/mars/util/EditorFont.java
package edu.missouristate.mars.util;

import edu.missouristate.mars.Globals;

import java.awt.*;
import java.util.Arrays;

/**
 * Specialized Font class designed to be used by both the
 * settings menu methods and the Settings class.
 *
 * @author Pete Sanderson
 * @version July 2007
 */
@SuppressWarnings("MagicConstant")
public class EditorFont {
    // Note: These are parallel arrays so corresponding elements must match up.
    private static final String[] styleStrings = {"Plain", "Bold", "Italic", "Bold + Italic"};
    private static final int[] styleConstants = {Font.PLAIN, Font.BOLD, Font.ITALIC, Font.BOLD | Font.ITALIC};
    public static final String DEFAULT_STYLE_STRING = styleStrings[0];
    public static final int DEFAULT_STYLE_INT = styleConstants[0];
    public static final int MIN_SIZE = 6;
    public static final int MAX_SIZE = 72;
    public static final int DEFAULT_SIZE = 12;
    /* Fonts in 3 categories that are common to major Java platforms: Win, Mac, Linux.
     *    Monospace: Courier New and Lucida Sans Typewriter
     *    Serif: Georgia, Times New Roman
     *    Sans Serif: Ariel, Verdana
     * This is according to lists published by www.codestyle.org.
     */
    private static final String[] allCommonFamilies = {"Arial", "Courier New", "Georgia", "Lucida Sans Typewriter", "Times New Roman", "Verdana"};

    /**
     * Obtain an array of common font family names.  These are guaranteed to
     * be available at runtime, as they were checked against the local
     * GraphicsEnvironment.
     *
     * @return Array of strings, each is a common and available font family name.
     */
    public static String[] getCommonFamilies() {
        return commonFamilies;
    }

    /**
     * Obtain an array of all available font family names.  These are guaranteed to
     * be available at runtime, as they come from the local GraphicsEnvironment.
     *
     * @return Array of strings, each is an available font family name.
     */
    public static String[] getAllFamilies() {
        return GraphicsEnvironment.getLocalGraphicsEnvironment().getAvailableFontFamilyNames();
    }

    /**
     * Get the array containing String values for font style names.
     */
    public static String[] getFontStyleStrings() {
        return styleStrings;
    }

    /**
     * Given a string that represents a font style, returns the
     * corresponding final int defined in Font: PLAIN, BOLD, ITALIC.  It
     * is not case-sensitive.
     *
     * @param style String representing the font style name
     * @return The int value of the corresponding Font style constant.  If the
     * string does not match any style name, returns Font.PLAIN.
     */
    public static int styleStringToStyleInt(String style) {
        String styleLower = style.toLowerCase();
        for (int i = 0; i < styleStrings.length; i++) {
            if (styleLower.equals(styleStrings[i].toLowerCase())) {
                return styleConstants[i];
            }
        }
        return DEFAULT_STYLE_INT;
    }

    /**
     * Given an int that represents a font style from the Font class,
     * returns the corresponding String.
     *
     * @param style Must be one of {@code Font.PLAIN}, {@code Font.BOLD}, or {@code Font.ITALIC}.
     * @return The String representation of that style.
     * If the parameter is not one of the above, returns "Plain".
     */
    public static String styleIntToStyleString(int style) {
        for (int i = 0; i < styleConstants.length; i++) {
            if (style == styleConstants[i]) {
                return styleStrings[i];
            }
        }
        return DEFAULT_STYLE_STRING;
    }

    /**
     * Given an int representing the font size, returns corresponding string.
     *
     * @param size Int representing size.
     * @return String value of parameter, unless it is less than MIN_SIZE (returns MIN_SIZE
     * as String) or greater than MAX_SIZE (returns MAX_SIZE as String).
     */
    public static String sizeIntToSizeString(int size) {
        int result = (size < MIN_SIZE) ? MIN_SIZE : Math.min(size, MAX_SIZE);
        return String.valueOf(result);
    }

    /**
     * Given a String representing font size, returns corresponding int.
     *
     * @param size String representing size.
     * @return int value of parameter, unless it is less than MIN_SIZE (returns
     * MIN_SIZE) or greater than MAX_SIZE (returns MAX_SIZE).  If the string
     * cannot be parsed as a decimal integer, it returns DEFAULT_SIZE.
     */
    public static int sizeStringToSizeInt(String size) {
        int result = DEFAULT_SIZE;
        try {
            result = Integer.parseInt(size);
        } catch (NumberFormatException ignored) {
        }
        return (result < MIN_SIZE) ? MIN_SIZE : Math.min(result, MAX_SIZE);
    }

    /**
     * Creates a new Font object based on the given String specifications.  This
     * is not the same as {@link java.awt.Font}'s constructor, which requires integer values for style and size.
     * It assures that defaults and size limits are applied when necessary.
     *
     * @param family String containing font family.
     * @param style  String containing font style.  A list of available styles can
     *               be obtained from getFontStyleStrings().  The default of styleStringToStyleInt()
     *               is substituted if necessary.
     * @param size   String containing font size.  The defaults and limits of
     *               sizeStringToSizeInt() are substituted if necessary.
     */
    public static Font createFontFromStringValues(String family, String style, String size) {
        return new Font(family, styleStringToStyleInt(style), sizeStringToSizeInt(size));
    }

    private static final String TAB_STRING = "\t";
    private static final char TAB_CHAR = '\t';
    private static final String SPACES = "                                                  ";

    /**
     * Handy utility to produce a string that substitutes spaces for all tab characters
     * in the given string.  The number of spaces generated is based on the position of
     * the tab character and the editor's current tab size setting.
     *
     * @param string The original string
     * @return New string in which spaces are substituted for tabs
     * @throws NullPointerException if string is null
     */
    public static String substituteSpacesForTabs(String string) {
        return substituteSpacesForTabs(string, Globals.getSettings().getEditorTabSize());
    }

    /**
     * Handy utility to produce a string that substitutes spaces for all tab characters
     * in the given string.  The number of spaces generated is based on the position of
     * the tab character and the specified tab size.
     *
     * @param string  The original string
     * @param tabSize The number of spaces each tab character represents
     * @return New string in which spaces are substituted for tabs
     * @throws NullPointerException if string is null
     */
    public static String substituteSpacesForTabs(String string, int tabSize) {
        if (!string.contains(TAB_STRING)) return string;
        StringBuilder result = new StringBuilder(string);
        for (int i = 0; i < result.length(); i++) {
            if (result.charAt(i) == TAB_CHAR) {
                result.replace(i, i + 1, SPACES.substring(0, tabSize - (i % tabSize)));
            }
        }
        return result.toString();
    }

    /*
     * We want to vet the above list against the actual available families and give
     * our client only those that are actually available.
     */
    private static final String[] commonFamilies = actualCommonFamilies();

    private static String[] actualCommonFamilies() {
        String[] result = new String[allCommonFamilies.length];
        String[] availableFamilies = GraphicsEnvironment.getLocalGraphicsEnvironment().getAvailableFontFamilyNames();
        Arrays.sort(availableFamilies); // not sure if necessary; is the list already alphabetical?
        int k = 0;
        for (String allCommonFamily : allCommonFamilies) {
            if (Arrays.binarySearch(availableFamilies, allCommonFamily) >= 0) {
                result[k++] = allCommonFamily;
            }
        }
        // If not all are found, creat a new array with only the ones that are.
        if (k < allCommonFamilies.length) {
            String[] temp = new String[k];
            System.arraycopy(result, 0, temp, 0, k);
            result = temp;
        }
        return result;
    }
}

File: src/main/java/edu/missouristate/mars/util/EditorFont.kt
/*
 * Copyright (c) 2003-2024, Pete Sanderson and Kenneth Vollmar
 * Copyright (c) 2024-present, Nicholas Hubbard
 *
 * Originally developed by Pete Sanderson (psanderson@otterbein.edu) and Kenneth Vollmar (kenvollmar@missouristate.edu)
 * Maintained by Nicholas Hubbard (nhubbard@users.noreply.github.com)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * 1. The above copyright notice and this permission notice shall be included in all copies or substantial portions of
 *    the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
 * WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
 * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package edu.missouristate.mars.util

import edu.missouristate.mars.Globals
import java.awt.Font
import java.awt.GraphicsEnvironment
import kotlin.math.max
import kotlin.math.min

/**
 * Specialized Font class designed to be used by both the settings menu methods and the Settings class.
 */
object EditorFont {
    enum class FontStyles(val styleName: String, val constant: Int) {
        PLAIN("Plain", Font.PLAIN),
        BOLD("Bold", Font.BOLD),
        ITALIC("Italic", Font.ITALIC),
        BOLD_ITALIC("Bold + Italic", Font.BOLD or Font.ITALIC);

        companion object {
            @JvmStatic val styleStrings = entries.map(FontStyles::styleName).toTypedArray()
            @JvmStatic val styleConstants = entries.map(FontStyles::constant).toTypedArray()

            fun fromName(name: String): FontStyles =
                entries.firstOrNull { it.styleName == name } ?: PLAIN

            fun fromConstant(constant: Int): FontStyles =
                entries.firstOrNull { it.constant == constant } ?: PLAIN
        }
    }

    @JvmStatic val DEFAULT_STYLE = FontStyles.PLAIN

    @Deprecated(
        "Use FontStyles enum instead.",
        ReplaceWith("DEFAULT_STYLE.stringValue"),
        DeprecationLevel.ERROR
    )
    @JvmStatic
    val DEFAULT_STYLE_STRING = FontStyles.PLAIN.styleName

    @Deprecated(
        "Use FontStyles enum directly instead.",
        ReplaceWith("DEFAULT_STYLE.constant"),
        DeprecationLevel.ERROR
    )
    @JvmStatic
    val DEFAULT_STYLE_INT = FontStyles.PLAIN.constant

    const val DEFAULT_SIZE = 12
    const val MIN_SIZE = 6
    const val MAX_SIZE = 72

    @JvmStatic val allCommonFamilies =
        arrayOf("Arial", "Courier New", "Georgia", "Lucida Sans Typewriter", "Times New Roman", "Verdana")

    /**
     * Obtain an array of all available font family names. These are guaranteed to be available at runtime, as they come
     * from the local GraphicsEnvironment.
     */
    @JvmStatic val allFamilies: Array<String> = GraphicsEnvironment.getLocalGraphicsEnvironment().availableFontFamilyNames

    @JvmStatic val fontStyleStrings: Array<String> = FontStyles.styleStrings

    @JvmStatic val commonFamilies = actualCommonFamilies()

    private const val TAB_STRING = "\t"
    private const val TAB_CHAR = '\t'
    private const val SPACES = "                                                  "

    /**
     * Given a string that represents a font style, returns the
     * corresponding final int defined in Font: PLAIN, BOLD, ITALIC.  It
     * is not case-sensitive.
     *
     * @param style String representing the font style name
     * @return The int value of the corresponding Font style constant.  If the
     * string does not match any style name, returns Font.PLAIN.
     */
    @JvmStatic
    @Deprecated(
        "Use FontStyles.fromName(style).constant instead.",
        ReplaceWith(
            "FontStyles.fromName(style).constant",
            "edu.missouristate.mars.util.KEditorFont.FontStyles"
        ),
        DeprecationLevel.ERROR
    )
    fun styleStringToStyleInt(style: String) = FontStyles.fromName(style).constant

    /**
     * Given an int that represents a font style from the Font class,
     * returns the corresponding String.
     *
     * @param style Must be one of {@code Font.PLAIN}, {@code Font.BOLD}, or {@code Font.ITALIC}.
     * @return The String representation of that style.
     * If the parameter is not one of the above, returns "Plain".
     */
    @JvmStatic
    @Deprecated(
        "Use KEditorFont.FontStyles.fromConstant(style).styleName instead.",
        ReplaceWith(
            "FontStyles.fromConstant(style).styleName",
            "edu.missouristate.mars.util.KEditorFont.FontStyles"
        ),
        DeprecationLevel.ERROR
    )
    fun styleIntToStyleString(style: Int) = FontStyles.fromConstant(style).styleName

    /**
     * Given an int representing the font size, returns corresponding string.
     *
     * @param size Int representing size.
     * @return String value of parameter, unless it is less than MIN_SIZE (returns MIN_SIZE
     * as String) or greater than MAX_SIZE (returns MAX_SIZE as String).
     */
    @JvmStatic
    fun sizeIntToSizeString(size: Int): String = max(min(size, MAX_SIZE), MIN_SIZE).toString()

    /**
     * Given a String representing font size, returns corresponding int.
     *
     * @param size String representing size.
     * @return int value of parameter, unless it is less than MIN_SIZE (returns
     * MIN_SIZE) or greater than MAX_SIZE (returns MAX_SIZE).  If the string
     * cannot be parsed as a decimal integer, it returns DEFAULT_SIZE.
     */
    @JvmStatic
    fun sizeStringToSizeInt(size: String) = max(min(size.toIntOrNull() ?: DEFAULT_SIZE, MAX_SIZE), MIN_SIZE)

    /**
     * Creates a new Font object based on the given String specifications.  This
     * is not the same as {@link java.awt.Font}'s constructor, which requires integer values for style and size.
     * It assures that defaults and size limits are applied when necessary.
     *
     * @param family String containing font family.
     * @param style  String containing font style.  A list of available styles can
     *               be obtained from getFontStyleStrings().  The default of styleStringToStyleInt()
     *               is substituted if necessary.
     * @param size   String containing font size.  The defaults and limits of
     *               sizeStringToSizeInt() are substituted if necessary.
     */
    @JvmStatic
    fun createFontFromStringValues(family: String, style: String, size: String): Font =
        Font(family, FontStyles.fromName(style).constant, sizeStringToSizeInt(size))

    /**
     * Handy utility to produce a string that substitutes spaces for all tab characters
     * in the given string.  The number of spaces generated is based on the position of
     * the tab character and the specified tab size.
     *
     * @param string  The original string
     * @param tabSize The number of spaces each tab character represents
     * @return New string in which spaces are substituted for tabs
     * @throws NullPointerException if string is null
     */
    @JvmStatic
    @JvmOverloads
    fun substituteSpacesForTabs(string: String, tabSize: Int = Globals.settings.getEditorTabSize()): String {
        if (!string.contains(TAB_STRING)) return string
        val result = StringBuilder(string)
        for (i in result.indices)
            if (result[i] == TAB_CHAR)
                result.replace(i, i + 1, SPACES.substring(0, tabSize - (i % tabSize)))
        return result.toString()
    }

    @JvmStatic
    private fun actualCommonFamilies(): Array<String> =
        allCommonFamilies.filter {
            it in GraphicsEnvironment.getLocalGraphicsEnvironment().availableFontFamilyNames
        }.toTypedArray()
}